### ISSUE TITLE
Advanced indexing slower than numpy

### ISSUE THREAD

**Issue by jeromerony** [pytorch/pytorch#14687]
## 🐛 Bug

Advanced indexing is slow.

## To Reproduce

```python
Python 3.7.0 (default, Oct  9 2018, 10:31:47) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.0.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch                                                                                                                                            

In [2]: torch.__version__                                                                                                                                       
Out[2]: '1.0.0.dev20181202'

In [3]: x = torch.randn(1024)                                                                                                                                   

In [4]: idx = torch.rand(1024) > 0.5                                                                                                                            

In [5]: %%timeit 
   ...: x[idx] 
   ...:                                                                                                                                                   
35 µs ± 63.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: x_np, idx_np = x.numpy(), idx.numpy()                                                                                                                   

In [7]: %%timeit 
   ...: x_np[idx_np] 
   ...:                                                                                                                                                 
2.36 µs ± 8.62 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

## Expected behavior

PyTorch should be as fast as numpy when doing advanced indexing of tensors that do not require grad.

## Environment

 - PyTorch Version (e.g., 1.0): 1.0.0.dev20181202
 - OS (e.g., Linux): Ubuntu 18.04.1 LTS
 - How you installed PyTorch (`conda`, `pip`, source): conda install pytorch-nightly -c pytorch
 - Python version: 3.7.0

**Comment by soumith** [comment]
Thanks for filing the issue @jeromerony . @colesbury can you take a look at this?

**Comment by colesbury** [comment]
The overhead of the nonzero call is dominating here. This is a pretty small input. If you increase the size to 1024x1024, then PyTorch is slightly faster than NumPy.

If someone wants to speed-up THTensor_(nonzero), then this case will speed up too.

**Comment by fmassa** [comment]
@colesbury just out of curiosity, does numpy call `nonzero` for masked indexing as well? I suppose so, given that it makes things simpler.

**Comment by colesbury** [comment]
@fmassa it looks like NumPy has an implementation for `nonzero` specifically for masked indexing:

https://github.com/numpy/numpy/blob/04c2f33843782cd27a16f919477b1b27c4b01e5a/numpy/core/src/multiarray/mapping.c#L2159

**Comment by colesbury** [comment]
@harryhan618 that's an issue, but it's not advanced indexing. see #5388

**Comment by ngimel** [comment]
Duplicate of #29973
